mlrun/feature_store/retrieval/base.py

```diff
@@ -0,0 +1,170 @@
+                self._append_drop_column(feature_set.spec.timestamp_key)
+            for key in feature_set.spec.entities.keys():
+                self._append_index(key)
+
+        return self._generate_vector(
+            entity_rows,
+            entity_timestamp_column,
+            feature_set_objects=feature_set_objects,
+            feature_set_fields=feature_set_fields,
+            start_time=start_time,
+            end_time=end_time,
+        )
+
+    def _write_to_target(self):
+        if self._target:
+            is_persistent_vector = self.vector.metadata.name is not None
+            if not self._target.path and not is_persistent_vector:
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "target path was not specified"
+                )
+            self._target.set_resource(self.vector)
+            size = self._target.write_dataframe(self._result_df)
+            if is_persistent_vector:
+                target_status = self._target.update_resource_status("ready", size=size)
+                logger.info(f"wrote target: {target_status}")
+                self.vector.save()
+
+    def _set_indexes(self, df):
+        if self._index_columns and not self._drop_indexes:
+
+            if df.index is None or df.index.name is None:
+                index_columns_missing = []
+                for index in self._index_columns:
+                    if index not in df.columns:
+                        index_columns_missing.append(index)
+                if not index_columns_missing:
+                    return df.set_index(self._index_columns)
+                else:
+                    logger.warn(
+                        f"Can't set index, not all index columns found: {index_columns_missing}. "
+                        f"It is possible that column was already indexed."
+                    )
+        return df
+
+    @abc.abstractmethod
+    def _generate_vector(
+        self,
+        entity_rows,
+        entity_timestamp_column,
+        feature_set_objects,
+        feature_set_fields,
+        start_time=None,
+        end_time=None,
+    ):
+        raise NotImplementedError("_generate_vector() operation not supported in class")
+
+    def merge(
+        self,
+        entity_df,
+        entity_timestamp_column: str,
+        featuresets: list,
+        featureset_dfs: list,
+    ):
+        """join the entities and feature set features into a result dataframe"""
+        merged_df = entity_df
+        if entity_df is None and featureset_dfs:
+            merged_df = featureset_dfs.pop(0)
+            featureset = featuresets.pop(0)
+            entity_timestamp_column = (
+                entity_timestamp_column or featureset.spec.timestamp_key
+            )
+
+        for featureset, featureset_df in zip(featuresets, featureset_dfs):
+            if featureset.spec.timestamp_key:
+                merge_func = self._asof_join
+            else:
+                merge_func = self._join
```

Comment:
Normally one would define stubs for these functions (`_asof_join` and `_join`) on the base class.

Reply:
added stubs